torch/distributed/distributed_c10d.py

```diff
@@ -4414,6 +4414,21 @@ def _new_group_with_tag(
             raise ValueError(
                 "MPI backend doesn't support use_local_synchronization=True"
             )
+        if backend == Backend.NCCL:
+            # When there is a device bound to the default group,
+            # 'ncclCommSplit' will be used instead of 'ncclCommInitRank'.
+            # The API 'ncclCommSplit' requires all the ranks from the parent
+            # communicator to participate.
+            # Ranks not part of this sub-group will call 'perform_nocolor_split'.
+            if (
+                is_initialized()
+                and _get_default_group().bound_device_id
+                and len(ranks) < _get_default_group().size()
+            ):
+                raise ValueError(
+                    "NCCL backend doesn't support use_local_synchronization=True "
+                    "in sub-groups when device_id is used with init_process_group"
+                )
```

Comment:
Hmm, does that mean the `use_local_synchronization` is less relevant when `device_id` is provided? Maybe in a future refactorization we can entirely separate the two?